docs/matrix_format_spec.md

```diff
@@ -0,0 +1,101 @@
+# Format
+
+The secondary analysis pipelines shall write expression matrices and associated
+tabular metadata using the zarr version 2 directory store format. The format is
+specified more precisely
+[here](https://zarr.readthedocs.io/en/stable/spec/v2.html), but in general zarr
+stores contain groups and arrays. Groups can contain other groups and arrays.
+Arrays are stored chunked and compressed. The file names and directory
+structure of the zarr store convey the group and chunk structure.
+
+# Groups and Arrays
+
+Expression values and metadata are stored in seven arrays. These arrays are
+stored within a single zarr group called `expression_matrix`.
+
+- expression - The expression values themselves. Rows correspond to cells and
+  columns to genes. These must represent raw counts. Any normalization is
+  deferred, but metadata that enables normalization should be included in the
+  metadata arrays defined below.
+  - shape: `(n_cells, n_genes)`
+  - compressor: `Blosc(cname='lz4', clevel=5, shuffle=SHUFFLE, blocksize=0)`
+  - chunk shape: `(10000, n_genes)`
```

Comment:
I think we should emphasize the optimal chunk size depends on many factors. First, size of a chunk depends on:
* dtype
* n_genes
* sparsity of the matrix
* compression algorithm
* and the number of cells in each chunk (here 10000)

The last item works as a knob to adjust the chunk size. The optimal chunk size depends on the storage option. Having too many small chunks means more frequent disk requests while having a few larger chunks means more [read delay](https://medium.com/@duhroach/gcs-read-performance-of-large-files-bd53cfca4410).

In my tests on GCS I was aiming for 2-4 MB chunk sizes, I am not sure what would be a good option for AWS. [This](https://github.com/zarr-developers/zarr/issues/86) might be relevant.

Reply:
Hey thanks Saman. Yeah you're right that there are a number of considerations for determining the optimal chunk size. I also think there are two different places where we have to determine that: 1) in the HCA's data storage system and 2) in the matrices that are delivered to users. Here we're talking pretty much only about (1), and there I'm mostly optimizing for what an AWS lambda can hold in its maximum 3GB of memory and 5 minute running time. It's a little hand-wavey, but 10000 cells x 30000 genes seems to work.

For (2) we actually have a lot more flexibility and diversity in use cases. The size of the chunks I'm producing with this spec are ~40MB, but we can do some more testing around what we can deliver to users.